torchmultimodal/modules/encoders/mdetr_text_encoder.py

```diff
@@ -0,0 +1,284 @@
+    taking in embeddings instead of input IDs.
+
+    Args:   embedding_dim (int): Number of features in the input.
+            num_encoder_layers  (int): Number of layers in the encoder.
+            num_attention_heads (int): Number of heads in multi-head attention.
+            ffn_dimension (Optional[int]): Dimension of feedforward network inside
+                attention layers. Defaults to 4 * embedding_dim
+            dropout (float): dropout value in each layer. Default: 0.1.
+            normalize_before (bool): Whether to do PreNorm in encoder layers.
+                Default: False
+            return_all_layers (bool) Whether to return all layers (or just the last
+                one). Default: False
+
+    Inputs: embeddings (Tensor): Tensor of embeddings of a batch of input IDs.
+            attention_mask (Optional[Tensor]) Batch attention mask returned from
+                tokenizer (applied as padding mask inside self-attention).
+                Default: None
+    """ ""
+
+    def __init__(
+        self,
+        embedding_dim: int,
+        num_encoder_layers: int,
+        num_attention_heads: int,
+        ffn_dimension: Optional[int] = None,
+        dropout: float = 0.1,
+        normalize_before: bool = False,
+        return_all_layers: bool = False,
+    ):
+        super().__init__()
+        ffn_dimension = ffn_dimension or 4 * embedding_dim
+        layer = torch.nn.TransformerEncoderLayer(
+            d_model=embedding_dim,
+            nhead=num_attention_heads,
+            dim_feedforward=ffn_dimension,
+            dropout=dropout,
+            activation="gelu",
+            batch_first=True,
+            norm_first=normalize_before,
+        )
+        self.layers = torch.nn.TransformerEncoder(
+            encoder_layer=layer, num_layers=num_encoder_layers
+        )
+        self.normalize_before = normalize_before
+        self.embedding_layer_norm = nn.LayerNorm(embedding_dim)
+        self.return_all_layers = return_all_layers
+        self.embedding_dim = embedding_dim
+
+    def _forward_return_all_layers(
+        self,
+        embeddings: Tensor,
+        attention_mask: Optional[Tensor] = None,
+    ) -> List[torch.Tensor]:
+        encoded = embeddings
+        states = [encoded]
+        for layer in self.layers.layers:
+            encoded = layer(encoded, src_key_padding_mask=attention_mask)
+            states.append(encoded)
+        if self.normalize_before:
+            for i, state in enumerate(states):
+                states[i] = self.embedding_layer_norm(state)
+        return states
+
+    def _forward_return_last_layer(
+        self,
+        embeddings: torch.Tensor,
+        attention_mask: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+        encoded = embeddings
+        for layer in self.layers.layers:
+            encoded = layer(encoded, src_key_padding_mask=attention_mask)
+        if self.normalize_before:
```

Comment:
isnt this normalize after?

Reply:
`normalize_before` return? 😄 . This is also how TorchText does it. Lmk if you want me to change it